python/cudf/cudf/core/frame.py

```diff
@@ -582,16 +574,35 @@ def equals(self, other, **kwargs):
         else:
             return self._index.equals(other._index)
 
-    def _explode(self, explode_column_num: int, index: Optional[cudf.Index]):
-        if index is not None:
-            explode_column_num += index.nlevels
-        res_tbl = libcudf.lists.explode_outer(
-            cudf._lib.table.Table(self._data, index=index), explode_column_num
-        )
+    def _explode(self, explode_column: Any, ignore_index: bool):
+        """Helper function for `explode` in Series and Dataframe.
+        if the designated column to explode is non-nested, a copy
+        of the frame is returned. Otherwise, if ignore_index is
+        set, the original index is not exploded and will use
+        a `RangeIndex` instead.
+        """
+        if not is_list_dtype(self._data[explode_column].dtype):
+            copy = self.copy()
+            if ignore_index:
+                copy._index = cudf.RangeIndex(copy._num_rows)
+            return copy
 
+        explode_column_num = self._column_names.index(explode_column)
+        if ignore_index:
+            tmp_index, self._index = self._index, None
+        elif self._index is not None:
+            explode_column_num += self._index.nlevels
+
+        res_tbl = libcudf.lists.explode_outer(self, explode_column_num)
         res = self.__class__._from_table(res_tbl)
-        if index is not None:
-            res.index.names = index.names
+
+        res._data.multiindex = self._data.multiindex
+        res._data._level_names = self._data._level_names
```

Comment:
Do we not have a helper function that handles all of the "gotchas" in this?

Reply:
Checked with @shwina, and can't think of any.. `._data` is constructed inside `from_unique_ptr`, which has no information about `multiindex` and `_level_names.

Reply:
I agree it's not ideal. One alternative I've considered in the past is a custom `ColumnMeta` type containing information about column names:

```python
class ColumnMeta:
    names: Tuple[Any]
    multiindex: bool
    level_names: Tuple[Any]
```

And we pass objects of that type as arguments to `._from_unique_ptr`:

```python
cdef from_unique_ptr(unique_ptr[table] c_tbl, ColumnMeta data_meta, ColumnMeta index_meta=None):
    pass
```

Maybe a bit out of scope for this PR though.

Reply:
Definitely out of scope for this PR, but I like it a lot.

Reply:
Is `ColumnMeta` owned and maintained by some class? Or is it a message interface that gets created on the fly and passes around classes?

Reply:
Good question -- I think we should have a separate discussion about this outside the context of this PR. Happy to sync offline and then raise an issue?